cirq-core/cirq/value/measurement_key.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional
+
+import dataclasses
+
+MEASUREMENT_KEY_SEPARATOR = ':'
+
+
+@dataclasses.dataclass(frozen=True)
+class MeasurementKey:
+    """A class representing a Measurement Key.
+
+    Wraps a string key and provides validation. If you just want the string measurement key, simply
+    call `str()` on this.
+    TODO: Change allow_nested_key to is_nested_key when the nested key callers start interacting
+        with `MeasurementKey` directly. That way we can have tighter validation and won't need to
+        store allow_nested_key, is_nested_key can just be derived from name itself. This will
+        happen in Phase 2a detailed at
+        https://tinyurl.com/structured-measurement-keys#heading=h.zafcj653k11m
```

Comment:
What is the use case for persisting `allow_nested_key` beyond the constructor at all? My intuition is that whether nested keys are allowed would depend on the calling method, not the key itself.

Reply:
Yes, I wanted to make it an `InitVar` but we need it in the serialized representation to ensure that we don't raise an error when we de-serialize. We can dummy-add it by parsing the key name but that would introduce some inconsistency between construction and serialization.

Reply:
Discussed offline - the deserialization case is a good argument for keeping this parameter.